Refactor Site.js to use async/await

jonahtanjz opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Related #809.
Refactor Site.js promise chaining to async await.

**Anything you'd like to highlight / discuss:**
Code style for functions that were declared and executed inside `.then` previously.

Example: 
```
.then((assets) => {
  ...
})
```

is refactored to:
```
return await (async (assets) => {
  ...
})(assets)
```
Is the current implementation acceptable or should I extract these functions to improve readability?

**Testing instructions:**
NIL
<!--
  Any special testing instructions in not including our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Refactor Site.js to use async await

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!--
  Prefix your pull request title with [WIP] if any of these are not complete yet

  Tick non-applicable items as well
-->

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -174,12 +174,13 @@ class Site {
-        .catch((err) => {
-          reject(new Error(`Failed to initialize site with given template with error: ${err.message}`));
-        });
+    return new Promise(async (resolve, reject) => {
```
> Is the current implementation acceptable or should I extract these functions to improve readability?

let's extract them : )

there are a few redundant wrapping promises before already, let's remove those as well (e.g. here we could make the `initSite` method async instead)

jonahtanjz replied:
Alright I've extracted them and remove redundant promises wrapping.

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -1063,70 +1096,73 @@ class Site {
-            travisTokenVar = _.isBoolean(travisTokenVar) ? 'GITHUB_TOKEN' : travisTokenVar;
-            if (!process.env[travisTokenVar]) {
-              reject(new Error(`The environment variable ${travisTokenVar} does not exist.`));
+    return new Promise(async (resolve, reject) => {
```
another here

jonahtanjz replied:
Updated.

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -706,80 +719,90 @@ class Site {
+      Promise.resolve('');
+    } else {
+      await Promise.all(removeFiles);
+      return logger.debug('Assets removed');
```
no `return` needed here, could remove the top branch as well since its now async

jonahtanjz replied:
Removed.

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -706,80 +731,98 @@ class Site {
-    return Promise.all(regeneratePagesBeingViewed)
-      .then(() => fs.remove(this.tempPath));
+    await Promise.all(regeneratePagesBeingViewed);
+    return fs.remove(this.tempPath);
```
`await`

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -706,80 +731,98 @@ class Site {
-          this.mapAddressablePagesToPages(this.addressablePages || [], this.getFavIconUrl());
+    try {
+      await this.removeAsset(removedPageFilePaths);
+      return await this.rebuildRequiredPages();
```
here too, and rejectHandler below can be awaited

jonahtanjz commented:
@ang-zeyu I've updated the PR accordingly and ready for review. Let me know if I've missed out anything or if there are any other changes to be made 😃

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -706,80 +731,98 @@ class Site {
-      })
-      .catch(error => Site.rejectHandler(error, [this.tempPath, this.outputPath]));
+    logger.warn('Rebuilding all pages...');
+    return this.buildSourceFiles();
```
here

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -952,32 +1003,41 @@ class Site {
 
-    return fs.copy(faCssSrcPath, faCssDestPath).then(() => fs.copy(faFontsSrcPath, faFontsDestPath));
+    await fs.copy(faCssSrcPath, faCssDestPath);
+    return fs.copy(faFontsSrcPath, faFontsDestPath);
```
here

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -952,32 +1003,41 @@ class Site {
+      await this.generatePagesThrottled(pagesToRegenerate);
+      await this.writeSiteData();
+      logger.info('Pages rebuilt');
+      return this.calculateBuildTimeForRegenerateAffectedPages(startTime);
```
here and rejectHandler below

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -855,24 +898,18 @@ class Site {
+      const pageGenerationQueue = pages.map(page => async () => {
+        try {
+          await page.generate(this.externalManager);
+          return Site.generateProgressBarStatus(progressBar, counter, pageGenerationQueue, pages, resolve);
```
no return needed

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -855,24 +898,18 @@ class Site {
           logger.error(err);
-          reject(new Error(`Error while generating ${page.sourcePath}`));
-        }));
+          return reject(new Error(`Error while generating ${page.sourcePath}`));
```
here too

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -706,80 +731,98 @@ class Site {
-      .catch(error => Site.rejectHandler(error, [])); // assets won't affect deletion
+          fs.copy(path.join(this.rootPath, asset), path.join(this.outputPath, asset))))(listOfAssets);
+      await Promise.all(assetsToCopy);
+      return logger.info('Assets built');
```
here and below `rejectHandler`

ang-zeyu reviewed (commented):
nice work. couple more suggestions for optimizing, mainly based on `async` methods implicitly returning promises https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

A couple of these comments may be outdated as I'm seeing the refresh button, so take these with a pinch of salt

jonahtanjz commented:
@ang-zeyu I've updated the code accordingly and ready for review. I've also remove some additional `promise.resolve()` in the code. Do let me know if I've missed out on anything.

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -225,16 +224,9 @@ class Site {
-        reject(error);
-      }
-    });
+  async listAssets(fileIgnore) {
```
`async` and `await` in its method call not necessary here

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -462,7 +462,7 @@ class Site {
    * Collects the paths to be traversed as addressable pages
    */
-  collectAddressablePages() {
+  async collectAddressablePages() {
```
`async` not necessary here and in its method calls

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -1168,14 +1232,16 @@ class Site {
   }
 
-  _setTimestampVariable() {
+  async _setTimestampVariable() {
```
`async` not necessary here

ang-zeyu reviewed (commented):
Sorry @jonahtanjz, in hindsight this really wasn't something suitable for a single code sprint.

Anyway, let's take the chance to polish up a couple more 🙂

jonahtanjz commented:
@ang-zeyu I've also updated the Site.test.js and removed `await` for method calls to `collectAddressablePages()` and remove one promise `rejects` check.

ang-zeyu reviewed (approved):
Thanks @jonahtanjz! This has been really long overdue

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -696,90 +713,107 @@ class Site {
+      this.mapAddressablePagesToPages(this.addressablePages || [], this.getFavIconUrl());
+
+      await this.rebuildPageBeingViewed(this.currentPageViewed);
+      await this.lazyBuildAllPagesNotViewed();
```
@jonahtanjz just noticed this, let's re-add the missing `return` termination sometime
